### ISSUE TITLE
Make `List.sort_uniq` more consistent

### ISSUE THREAD

**Issue by benjub** [ocaml/ocaml#13699]
The function `List.sort_uniq` currently removes duplicates in a hardly consistent way (see below).  I was going to propose to write that explicitly in its documentation, but actually, it is possible with minimal changes to make `List.sort_uniq` consistently remove the non-first duplicates (that is, keep only the first occurrences): only very few changes in the base cases of the current `List.sort_uniq` are necessary (details below).  One could simply make these changes, or, paralleling the current `List.stable_sort`, one could name that new function `List.stable_sort_uniq` (or `List.sort_keep_first`) and define `let sort_uniq = stable_sort_uniq`.  I can make a PR if agreed, or let a contributor do it.

#### Illustration that it may be dangerous to use the current `List.sort_uniq` with a non-tight comparison function
```
# let tag l = List.combine l (List.init (List.length l) Fun.id)
  let sortuniq l = List.sort_uniq (fun c1 c2 -> Int.compare (fst c1) (fst c2)) (tag l)
  let nines n = List.init n (Fun.const 9)
  let notlast, notfirst = (sortuniq (nines 2), sortuniq (nines 3));;
val tag : 'a list -> ('a * int) list = <fun>
val sortuniq : int list -> (int * int) list = <fun>
val nines : int -> int list = <fun>
val notlast : (int * int) list = [(9, 0)]
val notfirst : (int * int) list = [(9, 1)]
```
In these two examples, the kept duplicate is respectively not the last, and not the first (nor the last) in the original list.

#### Proposal
It's only an 8-character change from the current version (changing 6 characters would be enough, but changing the other two keeps the code consistent).  Here is the diff, with line numbering beginning at the beginning of the function definition:
```
$ diff sortuniqOLD.ml sortuniqNEW.ml
1c1
< let sort_uniq cmp l =
---
> let stable_sort_uniq cmp l =
36,37c36,37
<             let c = cmp x2 x3 in
<             if c = 0 then [x2] else if c < 0 then [x2; x3] else [x3; x2]
---
>             let c = cmp x1 x3 in
>             if c = 0 then [x1] else if c < 0 then [x1; x3] else [x3; x1]
76,77c76,77
<             let c = cmp x2 x3 in
<             if c = 0 then [x2] else if c > 0 then [x2; x3] else [x3; x2]
---
>             let c = cmp x1 x3 in
>             if c = 0 then [x1] else if c > 0 then [x1; x3] else [x3; x1]
```

#### Test
```
# let stabsortuniq l =
    stable_sort_uniq (fun c1 c2 -> Int.compare (fst c1) (fst c2)) (tag l)
  let correctly_sorted sortuniq_variant l =
    let li = tag l in
    let sorted = sortuniq_variant l in
    List.equal Int.equal (List.map fst sorted) (List.sort_uniq Int.compare l)
    &&
    let first_kept xi =
      let x = fst xi in
      List.assoc x sorted = List.assoc x li
    in
    List.for_all first_kept li
  let randlist maxlen =
    let len = Random.int maxlen in
    List.init len (fun _ -> len + Random.int len);;
val stabsortuniq : int list -> (int * int) List.t = <fun>
val correctly_sorted : (int list -> (int * int) list) -> int list -> bool = <fun>
val randlist : int -> int list = <fun>
# let () =
    for _ = 0 to 999 do
      let l = randlist 99 in
      if not (correctly_sorted stabsortuniq l) then print_string "OH NO ";
      if not (correctly_sorted sortuniq l) then print_string "oh no ";
    done;;
oh no oh no oh no oh no oh no oh no oh no oh no oh no oh no ... 
```

**Comment by jberdine** [comment]
See also https://github.com/ocaml/ocaml/pull/10238 for prior discussion.

**Comment by benjub** [comment]
> See also #10238 for prior discussion.

Sorry, I had missed that. As for why I chose to keep first elements: I am not attached to it, but I considered "intra-module" consistency before looking at other modules, and the "first" elements are generally chosen: List.find and variants, List.assoc and variants (a fact that I used in my above test).

BTW, and again for consistency: List.mli contains two occurrences of "leftmost" which should be changed to "first".

**Comment by jberdine** [comment]
No need to apologize! I just wanted to make you aware of some other considerations that have come up previously. I don't have strong views on which of duplicate elements to keep, I would just prefer something that is simple to state in the documentation. If you are interested, I would be happy if you wanted to commandeer that old PR. I just ran out of time to push it and closed it to tidy the repo.

Regarding terminology, I believe that historically "leftmost" was used instead of "first" due to the fact that the first element added to a list is in fact the "rightmost" / "last" one, which could be a source of confusion.

**Comment by gasche** [comment]
Re-reading the discussion in #10238, I think that it is fine to revisit this issue, and that we should keep the leftmost/first element instead of the rightmost/last element -- this is more natural for most use-cases. One issue in #10238 was that we wanted to use `List.sort_uniq` to implement `Map.of_list`, and `Map.of_list` needs to retain the last element; but now we have implemented `Map.of_list` without sorting elements at all, which is probably less efficient but makes this thorny aspect of the question go away.

TL;DR: okay for a PR that tweaks `sort_uniq` to consistently keep the leftmost/first element.
